mllib/src/main/scala/org/apache/spark/ml/tuning/CrossValidator.scala

```diff
@@ -146,25 +147,18 @@ class CrossValidator @Since("1.2.0") (@Since("1.4.0") override val uid: String)
       val validationDataset = sparkSession.createDataFrame(validation, schema).cache()
       logDebug(s"Train split $splitIndex with multiple sets of parameters.")
 
+      val completeFitCount = new AtomicInteger(0)
```

Comment:
My understanding of Scala futures may be off here, but this seems to change the behavior to me. Now, the unpersist operation will happen in one of the training threads, instead of asynchronously in its own thread. I'm not sure how much of an effect that will have. 

Why can't you just put all the logic in one map statement like below:

````scala
      // Fit models in a Future for training in parallel
      val modelFutures = epm.zipWithIndex.map { case (paramMap, paramIndex) =>
        Future[Model[_]] {
          val model = est.fit(trainingDataset, paramMap).asInstanceOf[Model[_]]

          if (collectSubModelsParam) {
            subModels.get(splitIndex)(paramIndex) = model
          }
          // TODO: duplicate evaluator to take extra params from input
          val metric = eval.evaluate(model.transform(validationDataset, paramMap))
          logDebug(s"Got metric $metric for model trained with $paramMap.")
          metric
        } (executionContext)
      }

      // Unpersist training data only when all models have trained
      Future.sequence[Model[_], Iterable](modelFutures)(implicitly, executionContext)
        .onComplete { _ => trainingDataset.unpersist() } (executionContext)
````

Reply:
We hope to unpersist training dataset once all fitting finished. But your idea here will postpone the unpersist time until all fitting & evaluation done. and your code should have the same effect with:
```
val modelFutures = ...
val foldMetrics = modelFutures.map(ThreadUtils.awaitResult(_, Duration.Inf))
trainingDataset.unpersist()
validationDataset.unpersist()
```

and, about what you said:
> "Now, the unpersist operation will happen in one of the training threads, instead of asynchronously in its own thread. "

What's the possible effect or impact ? `trainingDataset.unpersist()` itself is a async method and won't block. So will it have some effect ? I think it can be put in any thread safely.

Reply:
You can also turn a sequence of futures into a future then map on that single future and do the unpersist there.

Reply:
@holdenk Yes, that's what's done in current master code, but, if in this way, the future have to be split into a `modelFuture` and the following `metricFuture`, so it cannot avoid the issue that: `modelFuture` still holds the `model` computed which cause the memory issue.

Reply:
You can use futures to do this, you need to use a var for `modelFutures` then map on those futures to `Unit` collect those into a sequence and map on that to unpersist, but why go to the trouble. What's the concern with doing it in the final training thread. Why is this a change in behavior?

Reply:
@MrBago About what your said:
> You can use futures to do this, you need to use a var for modelFutures, then map on those futures to Unit, then collect those into a sequence, then map on that to unpersist, and also set modelFutures to null to release those references

Can you post some pseudo code so I can check whether it works fine and its peak memory occupation.
Thanks!

Reply:
I think this might work, but I think what you have in the PR now is better.
```
      // Fit models in a Future for training in parallel
      var modelFutures = epm.map { paramMap =>
        Future[Model[_]] {
          val model = est.fit(trainingDataset, paramMap)
          model.asInstanceOf[Model[_]]
        } (executionContext)
      }

      // Unpersist training data only when all models have trained
      val unitFutures = modelFutures.map{ _.map{ _ => () } (executionContext) }
      Future.sequence[Unit, Iterable](unitFutures)(implicitly, executionContext)
        .onComplete { _ => trainingDataset.unpersist() } (executionContext)

      // Evaluate models in a Future that will calulate a metric and allow model to be cleaned up
      val foldMetricFutures = modelFutures.zip(epm).map { case (modelFuture, paramMap) =>
        modelFuture.map { model =>
          // TODO: duplicate evaluator to take extra params from input
          val metric = eval.evaluate(model.transform(validationDataset, paramMap))
          logDebug(s"Got metric $metric for model trained with $paramMap.")
          metric
        } (executionContext)
      }
      modelFutures = null
```